frontend/benefit/shared/src/utils/dates.ts

```diff
@@ -50,31 +50,35 @@ export const getMaxEndDate = (
 export const validateFinnishDatePattern = (value = ''): boolean =>
   /^([1-9]|[12]\d|3[01])\.([1-9]|1[0-2])\.20\d{2}/.test(value);
 
-export const validateDateIsFromCurrentYearOnwards = (
-  value: string
-): boolean => {
-  if (!value || value.length < 8) return false;
+export const getDateFromDateString = (value: string): Date | null => {
+  if (!value || value.length < 8) {
+    return null;
+  }
 
   const isFinnishDate = validateFinnishDatePattern(value);
   const date = isFinnishDate
     ? parse(value, 'd.M.yyyy', new Date())
     : parseDate(value);
 
   if (!date || !date?.toJSON()) {
-    return false;
+    return null;
   }
+
+  return date;
+};
+
+export const validateDateIsFromCurrentYearOnwards = (
+  value: string
+): boolean => {
+  const date = getDateFromDateString(value);
+
   return date ? date >= startOfYear(new Date()) : false;
 };
 
 export const validateIsTodayOrPastDate = (value: string): boolean => {
-  if (!value || value.length < 8) return false;
+  const date = getDateFromDateString(value);
 
-  const isFinnishDate = validateFinnishDatePattern(value);
```

Comment:
This new solution indeed looks better, but there's some hidden knowledge that I didn't find a reference to from JIRA.

Removing this probably breaks the app when using keyboard date input and should generally be run in all `yup` schemas that have dates. Inputting patterns like `2222`  which then the field tries to parse results to a app crashing fatal error.

See (this is not the only issue, sometimes happens in a single date field too):
![crash date](https://github.com/City-of-Helsinki/yjdh/assets/5328394/07c554a4-3337-4aee-bf96-5e330ee880bb)

Reply:
Also, sometimes the date is string and sometimes it's a date, thus the toJSON(), I'm unsure where this behaviour comes from.

Reply:
Here's the original issue: 

https://helsinkisolutionoffice.atlassian.net/issues/HL-964?jql=project%20%3D%20%22YJDH%20-%20Helsinki-lis%C3%A4%22%20AND%20text%20~%20%22kentt%C3%A4%22

Reply:
I'll see what can be done for invalid inputs in the new functions I added. As for editing the existing functions, my intention was to deduplicate code in them; the identical preamble checks were moved into `getDateFromDateString` and `validateFinnishDatePattern` is also now inside that. I tried to be careful so that their behaviors would stay the same, but if that isn't the case, then I'd be happy to be proven wrong.

Reply:
@EmiliaMakelaVincit oh I missed that, looking at confusing green/red diffs only. In all honesty, we should have been addressing this earlier when the date issues were spotted and I have no intention on pointing fingers here 😬

Reply:
Fair enough :D 

I tried to break the date pickers for the alteration and didn't manage to do so. Might be because they're simpler than over on the application side (no initial values, no automatic fill of a field based on another field).